src/see.cpp

```diff
@@ -0,0 +1,278 @@
+//' @param interior.only,interior_only Logical scalar indicating whether the intersection must
+//' be within the spans of the two lines.
+//' 
+//' @returns A numeric vector with x- and y-coordinates defining the
+//' intersecting point between the two lines.
+//' 
+//' @source Weisstein, Eric W. "Line-Line Intersection.
+//' "From MathWorld--A Wolfram Web Resource.
+//' \url{http://mathworld.wolfram.com/Line-LineIntersection.html}
+//' @author David Sterratt
+//' @export
+// [[Rcpp::export]]
+NumericVector line_line_intersection_rcpp(
+    NumericVector P1,
+    NumericVector P2,
+    NumericVector P3,
+    NumericVector P4,
+    bool interior_only=false) {
+  
+  double dx1 = P1[0] - P2[0];
+  double dx2 = P3[0] - P4[0];
+  double dy1 = P1[1] - P2[1];
+  double dy2 = P3[1] - P4[1];
+  
+  // Determinant of diff matrix
+  double D = dx1 * dy2 - dy1 * dx2;
+  
+  // If lines are parallel
+  if (D == 0.0) {
+    NumericVector inf_res = {R_PosInf, R_PosInf};
+    return(inf_res);
+  }
+  
+  // Determinant of x and y matrix
+  double D1 = P1[0] * P2[1] - P1[1] * P2[0];
+  double D2 = P3[0] * P4[1] - P3[1] * P4[0];
+  
+  // Intersection
+  double X = (D1 * dx2 - dx1 * D2) / D;
+  double Y = (D1 * dy2 - dy1 * D2) / D;
+  
+  if (interior_only) {
+    // Is intersection within span of each line?
+    double lambda1 = -((X - P1[0]) * dx1 + (Y - P1[1]) * dy1)/(pow(dx1, 2) + pow(dy1, 2));
+    double lambda2 = -((X - P3[0]) * dx2 + (Y - P3[1]) * dy2)/(pow(dx2, 2) + pow(dy2, 2));
+    
+    if (!((lambda1 > 0.0) & (lambda1 < 1.0) & (lambda2 > 0.0) & (lambda2 < 1.0))) {
+      NumericVector na_res = {NA_REAL, NA_REAL};
+      return(na_res);
+    }
+  }
+  
+  NumericVector res = NumericVector::create(X, Y);
+  
+  return(res);
+}
+
+
+//' Goal in Sight
+//' 
+//' Checks whether a goal \code{P} can be seen from point \code{p}, or if it is
+//' occluded by \code{objects}.
+//'
+//' @param p_n,P_n Numeric vector with x- and y-coordinates.
+//' @param objects List containing a list for each object. An object has
+//' two length-two numeric vectors of x- and y-coordinates.
+//' 
+//' @returns \code{TRUE} if the goal is in sight, \code{FALSE} otherwise.
+//' @examples
+//' objects = list(
+//'   list(x = c(0.5, 0.5), y = c(0.5, 0.5))
+//' )
+//' 
+//' seesGoal_rcpp(c(0, 0), c(1, 1), objects)
+//' # FALSE
+//' 
+//' @export
+// [[Rcpp::export]]
+bool seesGoal_rcpp(NumericVector p_n, NumericVector P_n, List objects) {
+  int l = objects.length();
+  
+  // Indices for creating P3 and P4
+  NumericVector idx_P3_x = NumericVector::create(0, 0, 0, 1);
+  NumericVector idx_P3_y = NumericVector::create(0, 0, 1, 0);
+  NumericVector idx_P4_x = NumericVector::create(1, 0, 1, 1);
+  NumericVector idx_P4_y = NumericVector::create(0, 1, 1, 1); 
+  
+  for (int i = 0; i < l; ++i) {
+    
+    List object_i = objects[i];
+    NumericVector x = object_i["x"];
+    NumericVector y = object_i["y"];
+    
+    for (int j = 0; j < 4; ++j) {

```

Comment:
you are creating a numeric vector within a for loop, this means that for each iteration you will allocate a separate chunk of memory. This is a bit inefficient, compared to creating the vector beforehand with the right dimension and filling it using indices. You may want to rethink this approach if your benchmark shows slow downs.